naga/src/front/wgsl/mod.rs

```diff
@@ -32,18 +34,46 @@ impl Frontend {
     }
 
     pub fn parse(&mut self, source: &str) -> Result<crate::Module, ParseError> {
-        self.inner(source).map_err(|x| x.as_parse_error(source))
+        self.parse_to_ast(source)?.to_module(None)
     }
 
-    fn inner<'a>(&mut self, source: &'a str) -> Result<crate::Module, Error<'a>> {
-        let tu = self.parser.parse(source)?;
-        let index = index::Index::generate(&tu)?;
-        let module = Lowerer::new(&index).lower(&tu)?;
+    /// Two-step module conversion, can be used to compile with a "base module".
+    pub fn parse_to_ast<'a>(&mut self, source: &'a str) -> Result<ParsedWgsl<'a>, ParseError> {
+        self.inner_to_ast(source)
+            .map_err(|x| x.as_parse_error(source))
+    }
 
-        Ok(module)
+    fn inner_to_ast<'a>(&mut self, source: &'a str) -> Result<ParsedWgsl<'a>, Error<'a>> {
+        let translation_unit = self.parser.parse(source)?;
+        let index = index::Index::generate(&translation_unit)?;
+        Ok(ParsedWgsl {
+            source,
+            translation_unit,
+            index,
+        })
     }
 }
 
+pub struct ParsedWgsl<'a> {
+    source: &'a str,
+    translation_unit: TranslationUnit<'a>,
+    index: index::Index<'a>,
+}
```

Comment:
This intermediate struct lets us repeatedly compile a module with different base modules. That's something that naga_oil may want, hence me adding support for it.